raiden/tasks.py

```diff
@@ -281,442 +439,793 @@ def _run(self):  # noqa pylint: disable=method-hidden,too-many-locals
+                )
+
+                to_channel = to_assetmanager.get_channel_by_partner_address(
+                    to_mediated_transfer.sender
+                )
+
+                # now the secret can be revealed forward (`from_hop`)
+                from_assetmanager.handle_secret(identifier, secret)
+                to_assetmanager.handle_secret(identifier, secret)
+
+                self._wait_for_unlock_or_close(
+                    raiden,
+                    to_assetmanager,
+                    to_channel,
+                    to_mediated_transfer,
+                )
+
+                from_transfermanager.on_hashlock_result(hashlock, True)
+                self.done_result.set(True)
+
+    def send_and_wait_valid_state(self, raiden, path, from_asset_transfer,  # noqa
+                                  to_asset, to_amount):
+        """ Start the exchange by sending the first mediated transfer to the
+        taker and wait for mediated transfer for the exchagend asset.
+
+        This method will validate the messages received, discards the invalid
+        ones, and wait until a valid state is reached. The valid state is
+        reached when a mediated transfer for `to_asset` with `to_amount` tokens
+        and a SecretRequest from the taker are received.
+
+        Returns:
+            None: when the timeout was reached.
+            MediatedTransfer: when a valid state is reached.
+            RefundTransfer/TransferTimeout: when a invalid state is reached by
+                our partner.
+        """
+        # pylint: disable=too-many-arguments
 
-            log.error('Partner sent an invalid message. %s', repr(response))
+        next_hop = path[1]
+        taker_address = path[-1]  # taker_address and next_hop might be equal
+
+        # a valid state must have a secret request from the maker and a valid
+        # mediated transfer for the new asset
+        received_secretrequest = False
+        mediated_transfer = None
+
+        response_iterator = self._send_and_wait_time(
+            raiden,
+            from_asset_transfer.recipient,
+            from_asset_transfer,
+            raiden.config['msg_timeout'],
+        )
+
+        for response in response_iterator:
+
+            if response is None:
+                if log.isEnabledFor(logging.DEBUG):
+                    log.debug(
+                        'EXCHANGE TRANSFER TIMED OUT hashlock:%s',
+                        pex(from_asset_transfer.lock.hashlock),
+                    )
+
+                return None
+
+            # The MediatedTransfer might be from `next_hop` or most likely from
+            # a different node.
+            #
+            # The other participant must not use a direct transfer to finish
+            # the asset exchange, ignore it
+            if isinstance(response, MediatedTransfer) and response.asset == to_asset:
+                # XXX: allow multiple transfers to add up to the correct amount
```

Comment:
Truth is that I got used to it and don't recall why, I believe this was one of the default task tags in eclipse. In that case I didnt use TODO because multiple paths per transfer is a new feature that I don't even know if we will implement.

Regardless, if you have a suggestion list of task tags I will gladly follow it :)